base/lock.jl

```diff
@@ -186,7 +186,7 @@ Create a counting semaphore that allows at most `sem_size`
 acquires to be in use at any time.
 Each acquire must be matched with a release.
 """
-mutable struct Semaphore
+mutable struct Semaphore <: AbstractLock
```

Comment:
Using subtyping without clarifying `AbstractLock` contract does not seem to be a good idea. For example, does `islock(x)` imply `lock(x)` does not block (if other tasks don't do anything) or there is at least one task that has acquired `x` but has not released `x`?

If it were using subtyping for sharing the implementation, it may be a good idea to just duplicate the code (or use other mechanisms).

Reply:
We should just clarify the contract in this PR and merge it.
This PR has the right behavour all that is left is to write it down.

Reply:
Here's my attempt to clarify the properties of `islocked` and `trylock`: #44820